backend/controller/evaljob_controller.go

```diff
@@ -0,0 +1,568 @@
+	var pod = corev1.Pod{}
+	if err := r.Get(ctx, types.NamespacedName{Namespace: job.Namespace, Name: job.Name}, &pod); err != nil {
+		return nil, err
+	}
+	for _, ref := range pod.OwnerReferences {
+		if ref.APIVersion == job.APIVersion &&
+			ref.Kind == job.Kind &&
+			ref.Name == job.Name {
+
+			return &pod, nil
+		}
+	}
+	return nil, fmt.Errorf("pod doesn't have proper entry in the OwnerReferences")
+}
+
+func (r *EvalJobReconciler) deleteJobPod(ctx context.Context, job *lmevalservicev1beta1.EvalJob) error {
+	pod := corev1.Pod{
+		TypeMeta: v1.TypeMeta{
+			Kind:       "Pod",
+			APIVersion: "v1",
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name:      job.Status.PodName,
+			Namespace: job.Namespace,
+			OwnerReferences: []v1.OwnerReference{
+				{
+					APIVersion: job.APIVersion,
+					Kind:       job.Kind,
+					Name:       job.Name,
+				},
+			},
+		},
+	}
+	return r.Delete(ctx, &pod, &client.DeleteOptions{})
+}
+
+func (r *EvalJobReconciler) handleComplete(ctx context.Context, log logr.Logger, job *lmevalservicev1beta1.EvalJob) (ctrl.Result, error) {
+	if job.Status.CompleteTime == nil {
+		r.Recorder.Event(job, "Normal", "JobCompleted",
+			fmt.Sprintf("Tthe EvalJob %s in namespace %s has completed",
+				job.Name,
+				job.Namespace))
+		// TODO: final wrap up/clean up
+		current := v1.Now()
+		job.Status.CompleteTime = &current
+		if err := r.Status().Update(ctx, job); err != nil {
+			log.Error(err, "failed to update status for completion")
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *EvalJobReconciler) handleCancel(ctx context.Context, log logr.Logger, job *lmevalservicev1beta1.EvalJob) (ctrl.Result, error) {
+	// delete the pod and update the state to complete
+	if _, err := r.getPod(ctx, job); err != nil {
+		// pod is gone. update status
+		job.Status.State = lmevalservicev1beta1.CompleteJobState
+		job.Status.Reason = lmevalservicev1beta1.FailedReason
+		job.Status.Message = err.Error()
+	} else {
+		job.Status.State = lmevalservicev1beta1.CompleteJobState
+		job.Status.Reason = lmevalservicev1beta1.CancelledReason
+		if err := r.deleteJobPod(ctx, job); err != nil {
+			// leave the state as is and retry again
+			log.Error(err, "failed to delete pod. scheduled a retry after 10 seconds")
+			return ctrl.Result{Requeue: true, RequeueAfter: time.Second * 10}, err
+		}
+	}
+
+	err := r.Status().Update(ctx, job)
+	if err != nil {
+		log.Error(err, "failed to update status for cancellation")
+	}
+	r.Recorder.Event(job, "Normal", "Cancelled",
+		fmt.Sprintf("Tthe EvalJob %s in namespace %s has cancelled and changed its state to Complete",
+			job.Name,
+			job.Namespace))
+	return ctrl.Result{}, err
+}
+
+func (r *EvalJobReconciler) createPod(job *lmevalservicev1beta1.EvalJob) *corev1.Pod {
+	var allowPrivilegeEscalation = false
+	var runAsNonRootUser = true
+	var ownerRefController = true
+	var runAsUser int64 = 1001030000
```

Comment:
I _think_ we want to just let this setting be auto-assigned, right? We certainly want non-root, but I think the best practice is to not hard-code an explicit non-root UID and instead let a random (anonymous) one be assigned.

Reply:
yes, this would be auto-assigned on OpenShift. I forgot to remove it since my local dev env is using `kind`. The proper way to support both local dev env and OpenShift is:
- getting the uid range from the namespace's annotation
- specify an ID in that range or a random id if the annotation is not found (which is other k8s cluster)